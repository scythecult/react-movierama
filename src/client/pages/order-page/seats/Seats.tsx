import { DEFAULT_CANVAS_HEIGHT, DEFAULT_CANVAS_WIDTH } from '@/client/lib/constants/common';
import type { SeatData } from '@/client/lib/types/OrderPageData';
import { Seat } from './seat/Seat';
import styles from './styles.module.css';

export type SeatsProps = {
  seats: SeatData[];
  onClick?: (id: number) => void;
  canvasWidth: number;
  canvasHeight: number;
};

export const Seats = (props: SeatsProps) => {
  const { seats, canvasWidth = DEFAULT_CANVAS_WIDTH, canvasHeight = DEFAULT_CANVAS_HEIGHT, onClick } = props;

  const seatNodes = seats.map((seatNode) => {
    const { id } = seatNode;

    return <Seat key={id} {...seatNode} onClick={() => onClick?.(id)} />;
  });

  return (
    <div className={styles.canvas} style={{ width: canvasWidth, height: canvasHeight }}>
      {seatNodes}
    </div>
  );
};
