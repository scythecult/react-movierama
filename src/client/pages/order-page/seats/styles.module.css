.canvas {
    position: relative;
    width: 490px;
    height: 400px;
}