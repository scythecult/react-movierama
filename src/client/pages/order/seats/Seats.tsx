import type { SeatData } from '@/client/lib/types/OrderData';
import { Seat } from './seat/Seat';
import styles from './styles.module.css';

let rowKey = 100;

export type SeatsProps = {
  seats: SeatData[][];
  onClick?: (payload: SeatData) => void;
};

export const Seats = (props: SeatsProps) => {
  const { seats, onClick } = props;

  //   const selectedSeats = seatsData.reduce<string[]>((initial, seat) => {
  //     const selectedIds = seat.filter((item) => item.isSelected).map((item) => item.id);

  //     initial.push(...selectedIds);

  //     return initial;
  //   }, []);

  //   localStorage.setItem(STORED_SEATS_KEY, JSON.stringify(selectedSeats));
  // }, []);

  const seatNodes = seats.map((seat) => {
    const rows = [];

    for (let i = 0; i < seat.length; i++) {
      const { id, isSelected } = seat[i];

      rows.push(<Seat key={id} {...seat[i]} onClick={() => onClick?.({ ...seat[i], isSelected: !isSelected })} />);
    }

    return (
      <div key={rowKey++} className={styles.seatsRow}>
        {rows}
      </div>
    );
  });

  return <div className={styles.seats}>{seatNodes}</div>;
};
