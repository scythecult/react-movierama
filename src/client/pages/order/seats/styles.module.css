.seats {
  display: flex;
  flex-direction: column;
  gap: 6px;
}

.seatsRow {
  display: flex;
  gap: 6px;
}

.seatsRow > :nth-of-type(2) {
  margin-right: 18px;
}

.seatsRow > :nth-last-of-type(2) {
  margin-left: 18px;
}
