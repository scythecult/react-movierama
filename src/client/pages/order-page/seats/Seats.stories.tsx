import type { Meta, StoryFn } from '@storybook/react-vite';
import { CANVAS_SIZE, SEATS } from '../../../lib/utils/mocks';
import { Seats, type SeatsProps } from './Seats';

export default {
  title: 'Pages/Seats',
  component: Seats,
  args: {
    seats: SEATS,
    canvasWidth: CANVAS_SIZE.width,
    canvasHeight: CANVAS_SIZE.height,
  },
} satisfies Meta<SeatsProps>;

const Template: StoryFn<SeatsProps> = (props) => <Seats {...props} />;

export const Default = Template.bind(null);
