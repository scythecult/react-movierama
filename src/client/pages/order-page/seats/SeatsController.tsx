import { useAppStore } from '@/client/lib/contexts/app-store/useAppStore';
import { Seats } from './Seats';

export const SeatsController = () => {
  const seats = useAppStore((state) => state.seats);
  const { width, height } = useAppStore((state) => state.canvas);
  const setIsSelected = useAppStore((state) => state.setIsSelected);
  const updateCart = useAppStore((state) => state.updateCart);
  const calculateTotalPrice = useAppStore((state) => state.updateCartTotalPrice);

  const handleSeatClick = (id: number) => {
    setIsSelected(id);
    updateCart();
    calculateTotalPrice();
  };

  return <Seats seats={seats} canvasWidth={width} canvasHeight={height} onClick={handleSeatClick} />;
};
