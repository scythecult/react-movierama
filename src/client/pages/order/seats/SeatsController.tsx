import { useAppStore } from '@/client/lib/hooks/useAppStore';
import { Seats } from '@/client/pages/order/seats/Seats';

export const SeatsController = () => {
  const seats = useAppStore((state) => state.seats);
  const { width, height } = useAppStore((state) => state.canvas);
  const setIsSelected = useAppStore((state) => state.setIsSelected);
  const updateCart = useAppStore((state) => state.updateCart);
  const calculateTotalPrice = useAppStore((state) => state.updateCartTotalPrice);

  const handleSeatClick = (id: number) => {
    setIsSelected(id);
    updateCart();
    calculateTotalPrice();
  };

  return <Seats seats={seats} canvasWidth={width} canvasHeight={height} onClick={handleSeatClick} />;
};
