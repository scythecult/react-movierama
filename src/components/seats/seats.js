import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { toggleActive } from "../../features/seats/seats-slice";
import { Seat } from "../seat/seat";

import classes from "./seats.module.css";

let rowKey = 100;

const Seats = () => {
  const { seatsData } = useSelector((state) => state.seats);
  const dispatch = useDispatch();

  const seats = seatsData.map((seat) => {
    const cols = [];

    for (let i = 0; i < seat.length; i++) {
      const { id } = seat[i];

      cols.push(<Seat key={id} {...seat[i]} onClick={() => dispatch(toggleActive({ id }))} />);
    }

    return (
      <div key={rowKey++} className={classes.seats}>
        {cols}
      </div>
    );
  });

  return seats;
};

export { Seats };
