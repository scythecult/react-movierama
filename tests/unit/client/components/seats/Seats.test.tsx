import { render } from '@testing-library/react';
import { Seats } from '@/client/lib/components/seats/Seats';

describe('Seats', () => {
  test('should correspond default layout', () => {
    const result = render(<Seats />);

    expect(result.container).toMatchSnapshot();
  });
});
