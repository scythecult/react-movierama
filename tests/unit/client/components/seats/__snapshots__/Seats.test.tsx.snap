// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Seats > should correspond default layout 1`] = `
<div>
  <div
    class="_seats_fafe62"
  >
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74 _occupied_357f74"
      disabled=""
      type="button"
    />
    <button
      class="_seat_357f74 _occupied_357f74"
      disabled=""
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
  </div>
  <div
    class="_seats_fafe62"
  >
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
  </div>
  <div
    class="_seats_fafe62"
  >
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74 _occupied_357f74"
      disabled=""
      type="button"
    />
    <button
      class="_seat_357f74 _occupied_357f74"
      disabled=""
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
  </div>
  <div
    class="_seats_fafe62"
  >
    <button
      class="_seat_357f74 _occupied_357f74"
      disabled=""
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
  </div>
  <div
    class="_seats_fafe62"
  >
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74 _occupied_357f74"
      disabled=""
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
    <button
      class="_seat_357f74"
      type="button"
    />
  </div>
</div>
`;
